nixos/modules/hardware/fw-fanctrl.nix

```
@@ -0,0 +1,125 @@
+{
+  config,
+  lib,
+  pkgs,
+  ...
+}:
+
+let
+  cfg = config.hardware.fw-fanctrl;
+  defaultConfig = lib.importJSON "${pkgs.fw-fanctrl}/share/fw-fanctrl/config.json";
+in
+{
+  options.hardware.fw-fanctrl = {
+    enable = lib.mkEnableOption "Enable fw-fanctrl systemd service and install the needed packages.";
+
+    disableBatteryTempCheck = lib.mkOption {
+      type = lib.types.bool;
+      default = false;
+      description = ''
+        Disable checking battery temperature sensor
+      '';
+    };
+    config = lib.mkOption {
+      default = { };
+      description = ''
+        Config for the fw-fanctrl service
+      '';
+      type = lib.types.attrsOf (
+        lib.types.submodule (
+          { options, ... }:
+          {
+            defaultStrategy = lib.mkOption {
+              type = lib.types.str;
+              default = "";
+              description = "Default strategy to use";
+            };
+            strategyOnDischarging = lib.mkOption {
+              type = lib.types.str;
+              default = "";
+              description = "Default strategy on discharging";
+            };
+            strategies = lib.mkOption {
+              default = { };
+              type = lib.types.attrsOf (
+                lib.types.submodule (
+                  { options, name, ... }:
+                  {
+                    options = {
+                      fanSpeedUpdateFrequency = lib.mkOption {
+                        type = lib.types.ints.unsigned;
+                        default = 5;
+                        description = "How often the fan speed should be updated in seconds";
+                      };
+                      movingAverageInterval = lib.mkOption {
+                        type = lib.types.ints.unsigned;
+                        default = 25;
+                        description = "Interval (seconds) of the last temperatures to use to calculate the average temperature";
+                      };
+                      speedCurve = lib.mkOption {
+                        default = [ ];
+                        description = "How should the speed curve look like";
+                        type = lib.types.listOf (
+                          lib.types.submodule (
+                            { options, ... }:
+                            {
+                              options = {
+                                temp = lib.mkOption {
+                                  type = lib.types.int;
+                                  default = 0;
+                                  description = "Temperature at which the fan speed should be changed";
+                                };
+                                speed = lib.mkOption {
+                                  type = lib.types.ints.unsigned;
```

```suggestion
                                  type = lib.types.ints.between 0 100;
```
if this makes sense here